### PR TITLE
Support partial AppComponentFactory in Robolectric

### PR DESCRIPTION
**Pull request by utzcoz**
Support to use `AppComponentFactory` in Robolectric from SDK 28. This PR only supports `BroadcastReceiver` with custom `AppComponentFactory`. In Robolectric, developers use `Robolectric#setupService` or `Robolectric#buildService` to initialize `Service` instance, and these methods are static, so it's difficult to leverage custom `AppComponentFactory` to initialize `Service` instance with custom constructor. And I don't find proper usage scenarios to use custom `AppComponentFactory` to initialize `ContentProvider`, `Application`, `Activity` and `ClassLoader`. So I leave them as not implemented state.

Fixes: https://github.com/robolectric/robolectric/issues/7581.

### CODE REVIEW

**Comment by utzcoz** [comment]
@hoisie Could you help to review this PR? Thanks.
cc @ZacSweers for extra reviewing and testing.

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java [hunk 1]
```diff
@@ -684,14 +690,40 @@ private String createTempDir(String name) {
...
+      if (loadedApk != null && Build.VERSION.SDK_INT >= P) {
+        ClassLoader classLoader;
+        if (application.getBaseContext() != null) {
+          classLoader = application.getBaseContext().getClassLoader();
+        } else {
+          classLoader = loadedApk.getClassLoader();
+        }
+        // TODO utzcoz How about one receiver has multiple actions?
+        try {
+          broadcastReceiver =
```
Comment (utzcoz) [inline robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:713]: @hoisie IMO, it's not great enough now, and it needs feedback before merging.

Comment (utzcoz) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java]: I resolved it with trying to initialize the receiver with all actions until initializing it succeed.

**Comment by utzcoz** [comment]
@hoisie Could you help to give a look at this PR? Thanks.

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java [hunk 2]
```diff
@@ -684,14 +690,52 @@ private String createTempDir(String name) {
...
+  static void registerBroadcastReceivers(
+      Application application, AndroidManifest androidManifest, LoadedApk loadedApk) {
     for (BroadcastReceiverData receiver : androidManifest.getBroadcastReceivers()) {
       IntentFilter filter = new IntentFilter();
       for (String action : receiver.getActions()) {
         filter.addAction(action);
       }
       String receiverClassName = replaceLastDotWith$IfInnerStaticClass(receiver.getName());
-      application.registerReceiver((BroadcastReceiver) newInstanceOf(receiverClassName), filter);
+      BroadcastReceiver broadcastReceiver = null;
```
Comment (hoisie) [inline robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:701]: There may be a slightly simpler solution. Many months ago I did a quick prototype, and this also seems to work. It just involves instantiating things manually.

```java
  /**
   * Instantiate a new broadcast receiver, potentially using the {@link
   * android.app.AppComponentFactory} defined in the manifest.
   */
  private static BroadcastReceiver newBroadcastReceiverP(
      String receiverClassName, ApplicationInfo applicationInfo) {
    String appComponentFactoryName = applicationInfo.appComponentFactory;

    AppComponentFactory appComponentFactory = null;
    if (appComponentFactoryName != null) {
      appComponentFactoryName = replaceLastDotWith$IfInnerStaticClass(appComponentFactoryName);
      appComponentFactory = (AppComponentFactory) newInstanceOf(appComponentFactoryName);
    }
    if (appComponentFactory == null) {
      appComponentFactory = new AppComponentFactory();
    }
    try {
      return appComponentFactory.instantiateReceiver(
          Shadow.class.getClassLoader(), receiverClassName, null);
    } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
      return null;
    }
  }
  ```

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java [hunk 3]
```diff
@@ -403,6 +404,11 @@ private Package loadAppPackage_measured(Config config, AndroidManifest appManife
       Path packageFile = appManifest.getApkFile();
       parsedPackage = ShadowPackageParser.callParsePackage(packageFile);
     }
+    if (parsedPackage != null
```
Comment (hoisie) [inline robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:407]: Is this for legacy resources only? I think for binary resources applicationInfo.appComponentFactory should already be populated?

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java (hunk 2 above)
Comment (utzcoz) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:701]: Oh, it looks better. I will update it with this method.

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java (hunk 3 above)
Comment (utzcoz) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:407]: I remember we don't populate `applicationInfo.appComponentFactory` for binary resources. I wrote an article to analyze `AppComponentFactory` last weekend, and we need to parse `appComponentFactory` from `AndroidManifest.xml` and pass it to `ApplicationInfo` manually. And I don't find any existing code that parses `appComponentFactory` and passes it to `ParsedPacakge`'s `ApplicationInfo`. If I miss something please let me know.

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java (hunk 2 above)
Comment (utzcoz) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java]: I decide to keep the current implementation with initializing `AppComponentFactory` and storing it in `LoadedApk` to avoid unnecessary initialization by following AOSP's implementation.

**Comment by utzcoz** [comment]
@hoisie What about this PR? I also want to know whether it can be merged into 4.10.

**Comment by hoisie** [comment]
Yeah it'll definitely be merged before 4.10. Let me run some validation tests for it right now.

**Comment by hoisie** [comment]
When I ran this against some tests internally at Google I saw a non-trivial amount of failures. I'll take a closer look at a random sample and see what's going on.

**Comment by hoisie** [comment]
Just as an FYI, there will have to be some kind of way to enable or disable this feature (e.g. a system property, like `robolectric.enableAppComponentFactory`). I can see situations where BroadcastReceiver are initialized in a different way, and maybe that way causes tests to throw an Exception.

**Comment by utzcoz** [comment]
@hoisie Could you help to provide examples for these edge cases? I can improve the current implementation based on these examples.

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java [hunk 4]
```diff
@@ -682,16 +688,56 @@ private String createTempDir(String name) {
...
+      Application application,
+      LoadedApk loadedApk,
+      BroadcastReceiverData receiver) {
+    ClassLoader classLoader;
+    if (application.getBaseContext() != null) {
+      classLoader = application.getBaseContext().getClassLoader();
+    } else {
+      classLoader = loadedApk.getClassLoader();
+    }
+    for (String action : receiver.getActions()) {
```
Comment (hoisie) [inline robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:702]: One issue that I am seeing is  it is possible that receiver.getActions() is an empty list. Previously this would still get that receiver to be registered, but now it will not get registered.

Comment (utzcoz) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:702]: I think I can try to fix this issue this weekend.

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java [hunk 5]
```diff
@@ -682,16 +688,56 @@ private String createTempDir(String name) {
         .toString();
   }
 
+  private static BroadcastReceiver newBroadcastReceiverFromP(
+      String receiverClassName,
+      Application application,
+      LoadedApk loadedApk,
+      BroadcastReceiverData receiver) {
+    ClassLoader classLoader;
```
Comment (hoisie) [inline robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:696]: You don't need all of this classloader logic, there is only one classloader that works (the Robolectric one). You can just use `Shadow.class.getClassLoader()`.

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java [hunk 6]
```diff
@@ -682,16 +688,56 @@ private String createTempDir(String name) {
...
+  static void registerBroadcastReceivers(
+      Application application, AndroidManifest androidManifest, LoadedApk loadedApk) {
     for (BroadcastReceiverData receiver : androidManifest.getBroadcastReceivers()) {
       IntentFilter filter = new IntentFilter();
       for (String action : receiver.getActions()) {
         filter.addAction(action);
       }
       String receiverClassName = replaceLastDotWith$IfInnerStaticClass(receiver.getName());
-      application.registerReceiver((BroadcastReceiver) newInstanceOf(receiverClassName), filter);
+      if (loadedApk != null && Build.VERSION.SDK_INT >= P) {
```
Comment (hoisie) [inline robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:735]: `RuntimeEnvironment.getApiLevel()` is preferable to `Build.VERSION.SDK_INT`, because sadly it's common for people to modify the return value of `Build.VERSION.SDK_INT`.

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java [hunk 7]
```diff
@@ -682,16 +688,56 @@ private String createTempDir(String name) {
...
+    if (application.getBaseContext() != null) {
+      classLoader = application.getBaseContext().getClassLoader();
+    } else {
+      classLoader = loadedApk.getClassLoader();
+    }
+    for (String action : receiver.getActions()) {
+      try {
+        // If one BroadcastReceiver has multiple actions, we will try to load it with action one
+        // by one until we can initialize this BroadcastReceiver correctly.
+        Intent intent = new Intent(action);
```
Comment (hoisie) [inline robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:706]: Why not just pass null for the intent? Is this modeled after some logic in Android? Seems like AppComponentFactory should be able to handle null intents

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java (hunk 4 above)
Comment (hoisie) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:702]: Sounds good, I'll make this change and see if there are any other issues.

Comment (hoisie) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:702]: But yeah it's possible to define a broadcast receiver in a manifest without any intent filters, and then send an intent to the broadcast receiver class directly.

Comment (utzcoz) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:702]: Yep. It can be a test case for non-action receiver.

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java (hunk 6 above)
Comment (utzcoz) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:735]: Thanks for pointing it out. It's my first time to know this usage of `SDK_INT`.

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java (hunk 7 above)
Comment (utzcoz) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:706]: Oh, sorry for that. I forget it. The method signature of `AppComponentFactory` allows null intent. I will fix it with above issues together at this weekend.

Comment (hoisie) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:706]: Sounds good! I did some tests and it seems like a null intent works fine for this. Also we will not need a system property after all :)

**Comment by utzcoz** [comment]
@hoisie I applied above suggestions, PTAL.

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java [hunk 8]
```diff
@@ -692,16 +698,54 @@ private String createTempDir(String name) {
...
+    if (loadedApk == null || loadedApk.getAppFactory() == null) {
+      return (BroadcastReceiver) newInstanceOf(receiverClassName);
+    } else if (receiver.getActions() == null || receiver.getActions().isEmpty()) {
+      try {
+        return loadedApk.getAppFactory().instantiateReceiver(classLoader, receiverClassName, null);
+      } catch (InstantiationException | IllegalAccessException | ClassNotFoundException ignored) {
+        // Ignore these errors
+      }
+    } else {
+      for (String action : receiver.getActions()) {
```
Comment (hoisie) [inline robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:713]: I don't think it's really safe to instantiate arbitrary intents here. If there are some app-specific intents, they may require some specific data or extras that aren't being provided.

I think at this point you always `null` for the Intent.

I would recommend doing removing all the logic that iterates through the intents and tries to instantiate them one-by-one.

Comment (utzcoz) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:713]: Robolectric registers a `BroadcastReceiver` instance with the filter that contains all actions that this `BroadcastReceiver` has now. I consider to initialize multiple receiver instances with all actions, but it can't ensure that this `BroadcastReceiver` is instanitated with exact users data, for example they might pass meta data when start this `BroadcastReceiver`. So I decide to use a mid-balance method to try to instantiate this receiver instance by iterating all actions until `AppComponectFactory` accepts one and produce valid receiver instance to us.

Comment (hoisie) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:713]: Does't it seem like the use case for `intent` is in the case where the broadcast receiver is lazy-loaded? For instance, if you send an intent to some arbitrary string action (`MY_ACTION`), and the broadcast receiver for (`MY_ACTION`) is not instantiated, it can then be instantiated with the intent that was sent.

For instance, in this case, the receiver is being instantiated with a specific intent that was sent:
https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/app/ActivityThread.java;l=4298-4299;drc=c16c36030a21d6fc21901899d13477c31cc808a7;bpv=1;bpt=1

In AndroidTestEnvironment, there are no intents being sent, so I think we should always be passing `null` as the intent.

Is there some use case that I am missing?

Comment (utzcoz) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:713]: > In AndroidTestEnvironment, there are no intents being sent, so I think we should always be passing null as the intent.

Yep. The main reason that I use `Intent` is to avoid receiver's internal state is incorrect when initializing it with null as much as possible. Added tests contain this scenario that `AppComponentFactory` uses different constrcutor data to initialize receiver instance based on different `Intent` action.

Comment (utzcoz) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:713]: @hoisie Are there any progresses of this issue?

Comment (hoisie) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:713]: Yeah I still think we should be passing `null` here. This logic is executed at the start of each test , so instantiating arbitrary intents seems misleading.

Perhaps if the BroadcastReceiver instantiation was changed to be lazy we could route the correct intent to the AppCompotentFactory, but receiver instantiation is not lazy at the moment.

Comment (utzcoz) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:713]: > Yeah I still think we should be passing null here. This logic is executed at the start of each test , so instantiating receivers with arbitrary intents seems misleading.

Okay. I will update it with `null` tomorrow, and keep this behavior before we find extra issues from users. 

> Perhaps if the BroadcastReceiver instantiation was changed to be lazy we could route the correct intent to the AppCompotentFactory, but receiver instantiation is not lazy at the moment.

We can make receiver instantiation lazy, and I think it is more like the real AOSP logic.

Comment (utzcoz) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java]: @hoisie I updated the logic with `null`. You can check it again.

**Comment by utzcoz** [comment]
@hoisie I applied above suggestions, PTAL.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowLoadedApk.java [hunk 9]
```diff
@@ -23,6 +42,50 @@ public ClassLoader getSplitClassLoader(String splitName) throws NameNotFoundExce
...
+      android.app.AppComponentFactory factory =
+          (android.app.AppComponentFactory) newInstanceOf(fullQualifiedClassName);
+      if (factory == null) {
+        factory = new AppComponentFactory();
+      }
+      loadedApkReflector.setAppFactory(factory);
+    }
+  }
+
+  private static String replaceLastDotWith$IfInnerStaticClass(String receiverClassName) {
```
Comment (hoisie) [inline shadows/framework/src/main/java/org/robolectric/shadows/ShadowLoadedApk.java:68]: If you remove this method do the tests still pass?

Comment (hoisie) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowLoadedApk.java:68]: I removed this method from AndroidTestEnvironment and everything still works:

https://github.com/robolectric/robolectric/commit/40c16eaac4bbd63069ce8a41df681afc6b8561c3

Comment (utzcoz) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowLoadedApk.java:68]: I think it can work without this method. The reason that I copy it here is to keep the consistent with `AndroidTestEnvironment`. The critical part is to resolve the package name for class with the sytle `.TestClass`. I also will check whether we can remove it too because we use the single `ClassLoader` of the Robolectric.

**Comment by hoisie** [comment]
@utzcoz doing another round of tests on this, hopefully we can merge it tomorrow and then release 4.10.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowLoadedApk.java [hunk 10]
```diff
@@ -23,6 +42,50 @@ public ClassLoader getSplitClassLoader(String splitName) throws NameNotFoundExce
...
+        return;
+      }
+      String fullQualifiedClassName =
+          replaceLastDotWith$IfInnerStaticClass(
+              calculateFullQualifiedClassName(
+                  applicationInfo.appComponentFactory, applicationInfo.packageName));
+      android.app.AppComponentFactory factory =
+          (android.app.AppComponentFactory) newInstanceOf(fullQualifiedClassName);
+      if (factory == null) {
+        factory = new AppComponentFactory();
```
Comment (utzcoz) [inline shadows/framework/src/main/java/org/robolectric/shadows/ShadowLoadedApk.java:62]: @hoisie I think I need to remove `android.app.AppComponentFactory` import and use full qualified class name here.

**Comment by utzcoz** [comment]
@hoisie If you can wait, I can fix some issues of this PR tomorrow(Satuarday, East Eighth District TZ). Otherwise you can release 4.10, and we can port it to 4.10.1.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowLoadedApk.java (hunk 9 above)
Comment (hoisie) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowLoadedApk.java:68]: Yeah I confirmed that it works without this method, so it can be removed.

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java [hunk 11]
```diff
@@ -692,16 +697,42 @@ private String createTempDir(String name) {
...
 
+  private static BroadcastReceiver newBroadcastReceiverFromP(
+      String receiverClassName, LoadedApk loadedApk) {
+    ClassLoader classLoader = Shadow.class.getClassLoader();
+    if (loadedApk == null || loadedApk.getAppFactory() == null) {
+      return (BroadcastReceiver) newInstanceOf(receiverClassName);
+    } else {
+      try {
+        return loadedApk.getAppFactory().instantiateReceiver(classLoader, receiverClassName, null);
+      } catch (InstantiationException | IllegalAccessException | ClassNotFoundException ignored) {
```
Comment (hoisie) [inline robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java:708]: Can this be replaced with:
```
      } catch (ReflectiveOperationException e) {
        Logger.warn(
            "Failed to initialize receiver %s with AppComponentFactory %s: %s",
            receiverClassName, loadedApk.getAppFactory(), e);
        // Ignore these errors
      }
      ```

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowLoadedApk.java (hunk 9 above)
Comment (utzcoz) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowLoadedApk.java]: Updated, and let's check the CI result.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowLoadedApk.java (hunk 10 above)
Comment (utzcoz) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowLoadedApk.java]: Updated, and let's check the CI result.

File: robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java (hunk 11 above)
Comment (utzcoz) [reply robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java]: Updated. Let's check the CI result.

**Comment by utzcoz** [comment]
@hoisie Updated, PTAL.

**Review by hoisie — APPROVED** [review]
